proposals/downloads_get_file_hash.md

```diff
@@ -0,0 +1,152 @@
+# Proposal: `downloads.getFileHash()`
+
+**Summary**
+
+Allow extensions to programmatically query hashes of contents of downloaded files.
+
+**Document Metadata**
+
+**Author:** @bershanskyi
+
+**Sponsoring Browser:** Chromium
+
+**Contributors:** @Rob--W
+
+**Created:** 2024-04-08
+
+**Related Issues:** [WECG GitHub issue 401](https://github.com/w3c/webextensions/issues/401)
+
+## Motivation
+
+### Objective
+
+This API will enable extensions to privately and efficiently scan user
+downloads for "Known Unsafe" contents and validate integrity of "Known Safe"
+files without exposing the raw content of the files.
+
+#### Use Cases
+
+##### Filtering for unwanted downloads
+
+Many antivirus products detect previously seen "safe" and "unsafe" files by
+computing a hash of file content and checking it against a set of known hashes.
+These checks require very little effort on the client side, yet are powerful
+enough to limit the spread of known malware downloads since effecively every
+client gets "innoculated" against a particular download as soon as a single
+antivirus detects a malicious file.
+User Agents may store file hashes even after the original file is deleted to
+facilitate tracing of malvare exposure even if the file is deemed to be
+dangerous after the user initially downloads it.
+
+##### Download integrity assurance
+
+Files can get damaged while they are sored on file hosting server or NAS.
+While some file formats explicitly support integrity checks like embedded
+checksums or even cryptographic signatures, other files do not. This API
+would allow file managers or extensions to programmatically check the
+integrity of downloaded files if the website elects to provide them.
+
+### Known Consumers
+
+None at the moment.
+
+## Specification
+
+### Schema
+
+
+``` idl
+namespace downloads {
+    callback GetFileHashCallback = void(DOMString hash);
+    [supportsPromises, permissions="downloads.getFileHash"] static void getFileHash(
+        long downloadId,
+        [legalValues=("SHA-256")] DOMString algorithm,
+        GetFileHashCallback callback);
+}
+```
+
+``` idl
+namespace downloads {
+  enum FILE_HASH_ALGO {
+    "SHA-256"
+  }
+
+  getFileHash: (downloadId: number, algorithm?: downloads.FILE_HASH_ALGO) => Promise<string>;
+}
+```
+
+### New Permissions
+
+|      Permission Added     | Suggested Warning |
+| ------------------------- | ----------------- |
+| `"downloads.getFileHash"` | `"View cryptographic digests of downloads"` |
```

Comment:
Open questions:
- do we want a permission warning? If not, no more questions. If yes, continue.
- is this the permission warning that we'd like to settle on? It is quite technical and I am not sure if the average user would know what they are consenting to.
- do we want this API permission for any API?
- do we want to enable extensions with host permissions to finalUrl to get the hash without additional permission warnings?

Reply:
These are great questions, let's discuss them separately:
 > - do we want a permission warning?

Yes, I believe every permission should have a corresponding permission "message", even if it not a "warning". I have an ambition to (eventually) make every permission optional and adding a toggle in browser-provided UI.

> - is this the permission warning that we'd like to settle on?

Better proposals are welcome. I would like to add a "learn more" UI of some form eventually.

> - do we want this API permission for any API?

> - do we want to enable extensions with host permissions to `finalUrl` to get the hash without additional permission warnings?

Yes, but I intentionally left this permission relaxation out of this proposal and first implementation in Chromium. I would like to eventually move to host-based permission model. Specifically, if extension has a host permission for a site, it should be granted `"tabs"`, `"scripting"`, `"downloads"`, `"favicon"`, and many other permissions for that site specifically.

Reply:
Per WECG meeting minutes, we removed the permission warning and added requirement for host access to `DownloadItem.finalURL`. Should we mark this as resolved?